pkg/buildkit/worker/leasemanager.go

```diff
@@ -0,0 +1,345 @@
+	if err != nil {
+		return fmt.Errorf("cannot extract pod config: %w", err)
+	}
+
+	pac.WithAnnotations(map[string]string{
+		expiryTimeAnnotation: time.Now().Add(m.podMaxIdleTime).Format(time.RFC3339),
+	})
+	delete(pac.Annotations, leasedAnnotation)
+	delete(pac.Annotations, managerIDAnnotation)
+
+	m.log.Info("Applying pod metadata changes", "annotations", pac.Annotations)
+	if pod, err = m.podClient.Apply(ctx, pac, metav1.ApplyOptions{FieldManager: fieldManagerName}); err != nil {
+		return fmt.Errorf("cannot update pod metadata: %w", err)
+	}
+
+	go func() {
+		m.ptch <- podTransport{pod: pod, err: nil}
+	}()
+
+	return nil
+}
+
+func (m *podLeaseManager) nextAvailablePod(ctx context.Context) (*corev1.Pod, error) {
+	select {
+	case pt := <-m.ptch:
+		return pt.pod, pt.err
+	default:
+		go func(ctx context.Context) {
+			pod, err := m.getUnleasedOrScale(ctx)
+			m.ptch <- podTransport{pod: pod, err: err}
+		}(ctx)
+	}
+
+	res := <-m.ptch
+	return res.pod, res.err
+}
+
+func (m *podLeaseManager) getUnleasedOrScale(ctx context.Context) (*corev1.Pod, error) {
+	m.log.Info("Checking for unleased pods")
+	pod, err := m.getUnleasedPod(ctx)
+	if err == nil {
+		return pod, err
+	}
+
+	if !errors.Is(err, errNoUnleasedPods) {
+		return nil, fmt.Errorf("unexpected leasing error: %w", err)
+	}
+
+	m.log.Info("Scaling cluster, no unleased pods found")
+	if err = m.scaler.Scale(ctx, 1); err != nil {
+		return nil, err
+	}
+
+	m.log.Info("No unleased pods found, scaling workload")
+	if pod, err = m.getUnleasedPod(ctx); err != nil {
+		return nil, fmt.Errorf("unable to acquire lease after scaling: %w", err)
+	}
+
+	return pod, nil
+}
+
+func (m *podLeaseManager) getUnleasedPod(ctx context.Context) (*corev1.Pod, error) {
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range list.Items {
+		if _, ok := pod.Annotations[leasedAnnotation]; !ok {
+			m.log.Info("Found unleased pod", "name", pod.Name, "namespace", pod.Namespace)
+			return &pod, nil
+		}
+	}
+
+	m.log.Info("No unleased pods found")
+	return nil, errNoUnleasedPods
+}
+
+func (m *podLeaseManager) buildURL(ctx context.Context, podName string) (string, error) {
+	m.log.Info("Querying for endpoints", "name", m.serviceName, "namespace", m.namespace)
+	endpoints, err := m.endpointsClient.Get(ctx, m.serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	var host string
+
+addressScan:
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef.Name == podName {
+				for _, port := range subset.Ports {
+					if port.Port == m.servicePort {
+						host = strings.Join([]string{address.Hostname, endpoints.Name, endpoints.Namespace}, ".")
+						m.log.Info("Found eligible endpoint address", "host", host)
+
+						break addressScan
```

Comment:
This just screams separate method that takes endpoints and and has a `return host` to me. I had to lookup label breaks again.